File: bindings/csharp/Sop.Examples/BtreeBatched.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace Sop.Examples
{
    public static class BtreeBatched
    {
        public static void Run()
        {
            Console.WriteLine("--- Batched B-Tree Operations Demo ---");

            using var ctx = new Context();
            string dbPath = "data/batched_demo_db";
            if (Directory.Exists(dbPath)) Directory.Delete(dbPath, true);

            var db = new Database(new DatabaseOptions
            {
                StoresFolders = new List<string> { dbPath },
                Type = (int)DatabaseType.Standalone
            });

            // 1. Batched Add
            Console.WriteLine("\n1. Batched Add (100 items)...");
            using (var trans = db.BeginTransaction(ctx))
            {
                var btree = db.NewBtree<string, string>(ctx, "batched_btree", trans);
                
                var items = new List<Item<string, string>>();
                for (int i = 0; i < 100; i++)
                {
                    items.Add(new Item<string, string> { Key = $"key_{i}", Value = $"value_{i}" });
                }

                btree.Add(ctx, items);
                trans.Commit();
            }
            Console.WriteLine("Committed.");

            // 2. Batched Update
            Console.WriteLine("\n2. Batched Update (100 items)...");
            using (var trans = db.BeginTransaction(ctx))
            {
                var btree = db.OpenBtree<string, string>(ctx, "batched_btree", trans);
                
                var items = new List<Item<string, string>>();
                for (int i = 0; i < 100; i++)
                {
                    items.Add(new Item<string, string> { Key = $"key_{i}", Value = $"updated_value_{i}" });
                }

                btree.Update(ctx, items);
                trans.Commit();
            }
            Console.WriteLine("Committed.");

            // Verify Update
            using (var trans = db.BeginTransaction(ctx, TransactionMode.ForReading))
            {
                var btree = db.OpenBtree<string, string>(ctx, "batched_btree", trans);
                var item = btree.GetValues(ctx, new List<Item<string, string>> { new Item<string, string> { Key = "key_50" } });
                Console.WriteLine($"Verified key_50 value: {item[0].Value}");
                trans.Commit();
            }

            // 3. Batched Remove
            Console.WriteLine("\n3. Batched Remove (100 items)...");
            using (var trans = db.BeginTransaction(ctx))
            {
                var btree = db.OpenBtree<string, string>(ctx, "batched_btree", trans);
                
                var keys = new List<string>();
                for (int i = 0; i < 100; i++)
                {
                    keys.Add($"key_{i}");
                }

                btree.Remove(ctx, keys);
                trans.Commit();
            }
            Console.WriteLine("Committed.");

            // Verify Remove
            using (var trans = db.BeginTransaction(ctx, TransactionMode.ForReading))
            {
                var btree = db.OpenBtree<string, string>(ctx, "batched_btree", trans);
                long count = btree.Count();
                Console.WriteLine($"Verified count: {count}");
                trans.Commit();
            }

            Console.WriteLine("--- End of Batched Demo ---");
        }
    }
}

File: bindings/java/src/main/java/com/sharedcode/sop/examples/BTreeBatched.java
package com.sharedcode.sop.examples;

import com.sharedcode.sop.*;
import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BTreeBatched {
    public static void run() {
        System.out.println("--- Batched B-Tree Operations Demo ---");

        try (Context ctx = new Context()) {
            String dbPath = "data/batched_demo_db";
            File dbDir = new File(dbPath);
            if (dbDir.exists()) {
                deleteDirectory(dbDir);
            }

            DatabaseOptions dbOpts = new DatabaseOptions();
            dbOpts.stores_folders = Collections.singletonList(dbPath);
            dbOpts.type = DatabaseType.Standalone;
            
            Database db = new Database(dbOpts);

            // 1. Batched Add
            System.out.println("\n1. Batched Add (100 items)...");
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<String, String> btree = BTree.create(ctx, "batched_btree", trans, null, String.class, String.class);
                
                List<Item<String, String>> items = new ArrayList<>();
                for (int i = 0; i < 100; i++) {
                    items.add(new Item<>("key_" + i, "value_" + i));
                }

                btree.add(items);
                trans.commit();
            }
            System.out.println("Committed.");

            // 2. Batched Update
            System.out.println("\n2. Batched Update (100 items)...");
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<String, String> btree = db.openBtree(ctx, "batched_btree", trans, String.class, String.class);
                
                List<Item<String, String>> items = new ArrayList<>();
                for (int i = 0; i < 100; i++) {
                    items.add(new Item<>("key_" + i, "updated_value_" + i));
                }

                btree.update(items);
                trans.commit();
            }
            System.out.println("Committed.");

            // Verify Update
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<String, String> btree = db.openBtree(ctx, "batched_btree", trans, String.class, String.class);
                List<Item<String, String>> items = btree.getValues(Collections.singletonList(new Item<>("key_50", null)));
                if (items != null && !items.isEmpty()) {
                    System.out.println("Verified key_50 value: " + items.get(0).value);
                }
                trans.commit();
            }

            // 3. Batched Remove
            System.out.println("\n3. Batched Remove (100 items)...");
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<String, String> btree = db.openBtree(ctx, "batched_btree", trans, String.class, String.class);
                
                List<String> keys = new ArrayList<>();
                for (int i = 0; i < 100; i++) {
                    keys.add("key_" + i);
                }

                btree.remove(keys);
                trans.commit();
            }
            System.out.println("Committed.");

            // Verify Remove
            try (Transaction trans = db.beginTransaction(ctx)) {
                BTree<String, String> btree = db.openBtree(ctx, "batched_btree", trans, String.class, String.class);
                long count = btree.count();
                System.out.println("Verified count: " + count);
                trans.commit();
            }

            System.out.println("--- End of Batched Demo ---");
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static void deleteDirectory(File directory) {
        File[] files = directory.listFiles();
        if (files != null) {
            for (File file : files) {
                deleteDirectory(file);
            }
        }
        directory.delete();
    }

    public static void main(String[] args) {
        run();
    }
}
